src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.cpp

```diff
@@ -0,0 +1,468 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// This code compiles, but was never executed, be careful!
+
+#include "QubitManagerRestrictedReuse.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+
+//
+// QubitListInSharedArray
+//
+
+QubitListInSharedArray::QubitListInSharedArray(int minElement, int maxElement, int* sharedQubitStatusArray)
+{
+    // We are not storing pointer to shared array because it can be reallocated.
+    // Indexes and special values remain the same on such reallocations.
+    FailIf(minElement > maxElement || minElement < 0 || maxElement == INT_MAX,
+        "Incorrect boundaries in the linked list initialization.");
+
+    for (int i = minElement; i < maxElement; i++) {
+        sharedQubitStatusArray[i] = i + 1;
+    }
+    sharedQubitStatusArray[maxElement] = NoneMarker;
+    firstElement = minElement;
+    lastElement = maxElement;
+}
+
+bool QubitListInSharedArray::IsEmpty()
+{
+    return firstElement == NoneMarker;
+}
```

Comment:
Nit: Consider using `const` member functions.
> bool QubitListInSharedArray::IsEmpty() **const**

See "const Member Functions" in "Item 3: Use `const` whenever possible" either [here](http://wavelino.coffeecup.com/pdf/Effective%20Cshapshap.pdf) or [here](https://www.amazon.com/gp/product/0321334876?ie=UTF8&tag=aristeia.com-20&linkCode=as2&camp=1789&creative=9325&creativeASIN=0321334876).

Reply:
I made a few members const that seem to have no reason to modify containing classes. Although that required QubitToId to become const as well. And that's a virtual function. It doesn't modify containing class in my implementation, but I can imagine some overrides may actually need it. OK, let's be more restrictive for now.